File: src/sentence_builder.py
"""SentenceBuilder - Assembles individual letter predictions into words and sentences."""

import time
from typing import Callable, Optional


class SentenceBuilder:
    """Converts a stream of letter predictions into words and sentences."""

    def __init__(
        self,
        stability_ms: int = 500,
        space_no_hand_ms: int = 2000,
        enter_no_hand_ms: int = 4000,
    ):
        """Initialize timing thresholds and state.

        Args:
            stability_ms: Time a letter must hold before committing.
            space_no_hand_ms: No-hand duration to insert a space.
            enter_no_hand_ms: No-hand duration to complete a sentence.
        """
        # Timing thresholds
        self.stability_threshold = stability_ms
        self.space_no_hand_ms = space_no_hand_ms
        self.enter_no_hand_ms = enter_no_hand_ms

        # State
        self.current_letter: Optional[str] = None
        self.letter_start_time: float = 0
        self.letter_stable: bool = False
        self.current_word: str = ''
        self.words: list[str] = []
        self.sentences: list[str] = []
        self.last_hand_time: float = time.time()
        self.last_committed_letter: str = ''

        # Optional callbacks
        self.on_letter_commit: Optional[Callable] = None
        self.on_word_complete: Optional[Callable] = None
        self.on_sentence_complete: Optional[Callable] = None
        self.on_gesture: Optional[Callable] = None

    def feed(
        self,
        prediction: Optional[dict],
        hand_detected: bool,
        gesture_state: Optional[str] = None,
    ) -> None:
        """Called every frame with a prediction result.

        Args:
            prediction: Dict with 'letter' and 'confidence', or None.
            hand_detected: Whether a hand is detected in the frame.
            gesture_state: One of "fist", "palm", "signing", "none", or None.
        """
        now = time.time()

        if not hand_detected:
            no_hand_duration = (now - self.last_hand_time) * 1000  # ms
            if no_hand_duration >= self.enter_no_hand_ms and self.current_word:
                self.complete_sentence()
            elif no_hand_duration >= self.space_no_hand_ms and self.current_word:
                self.insert_space()
            return

        self.last_hand_time = now

        # Gesture overrides
        if gesture_state == 'fist':
            self.insert_space()
            if self.on_gesture:
                self.on_gesture('space')
            return
        if gesture_state == 'palm':
            self.complete_sentence()
            if self.on_gesture:
                self.on_gesture('enter')
            return

        # Letter stability check
        if prediction and prediction.get('confidence', 0) > 0.5:
            letter = prediction['letter']
            if letter == self.current_letter:
                if not self.letter_stable:
                    elapsed = (now - self.letter_start_time) * 1000
                    if elapsed >= self.stability_threshold:
                        self.letter_stable = True
                        self._commit_letter(letter)
            else:
                self.current_letter = letter
                self.letter_start_time = now
                self.letter_stable = False

    def _commit_letter(self, letter: str) -> None:
        if letter == self.last_committed_letter:
            return  # dedup
        self.current_word += letter
        self.last_committed_letter = letter
        if self.on_letter_commit:
            self.on_letter_commit(letter, self.current_word)

    def insert_space(self) -> None:
        """Finish the current word and start a new one."""
        if not self.current_word:
            return
        self.words.append(self.current_word)
        if self.on_word_complete:
            self.on_word_complete(self.current_word, self.words)
        self.current_word = ''
        self.last_committed_letter = ''
        self.current_letter = None

    def complete_sentence(self) -> None:
        """Finish the current sentence from accumulated words."""
        if self.current_word:
            self.insert_space()
        if not self.words:
            return
        sentence = ' '.join(self.words)
        self.sentences.append(sentence)
        if self.on_sentence_complete:
            self.on_sentence_complete(sentence, self.sentences)
        self.words = []

    def manual_space(self) -> None:
        """Manual space trigger (for UI buttons)."""
        self.insert_space()

    def manual_enter(self) -> None:
        """Manual enter trigger (for UI buttons)."""
        self.complete_sentence()

    def clear_all(self) -> None:
        """Reset all state."""
        self.current_word = ''
        self.words = []
        self.last_committed_letter = ''
        self.current_letter = None

    def get_state(self) -> dict:
        """Return current builder state as a dict."""
        return {
            'current_word': self.current_word,
            'words': list(self.words),
            'sentences': list(self.sentences),
            'building_sentence': (
                ' '.join([*self.words, self.current_word])
                if self.current_word
                else ' '.join(self.words)
            ),
        }

File: meeting/static/js/sentence-builder.js
/**
 * SentenceBuilder - Assembles letter predictions into words and sentences.
 * JavaScript mirror of the Python SentenceBuilder.
 * Used by web/test.html and meeting/static/js/meeting.js.
 */
class SentenceBuilder {
    constructor(options = {}) {
        // Timing thresholds (milliseconds)
        this.stabilityThreshold = options.stabilityMs || 500;
        this.spaceNoHandMs = options.spaceNoHandMs || 2000;
        this.enterNoHandMs = options.enterNoHandMs || 4000;
        this.gestureHoldMs = options.gestureHoldMs || 2000;

        // State
        this.currentLetter = null;
        this.letterStartTime = 0;
        this.letterStable = false;
        this.currentWord = '';
        this.words = [];
        this.sentences = [];
        this.lastHandTime = Date.now();
        this.lastCommittedLetter = '';

        // Gesture tracking
        this.currentGesture = null;
        this.gestureStartTime = 0;
        this.gestureTriggered = false;

        // Callbacks
        this.onLetterCommit = options.onLetterCommit || (() => {});
        this.onWordComplete = options.onWordComplete || (() => {});
        this.onSentenceComplete = options.onSentenceComplete || (() => {});
        this.onGesture = options.onGesture || (() => {});
        this.onBackspace = options.onBackspace || (() => {});
        this.onDeleteWord = options.onDeleteWord || (() => {});
    }

    /**
     * Called every frame with prediction result.
     * @param {Object|null} prediction - {letter, confidence} or null
     * @param {boolean} handDetected - whether a hand is detected
     * @param {string|null} gestureState - "fist" | "palm" | "signing" | null
     */
    feed(prediction, handDetected, gestureState) {
        const now = Date.now();

        console.log(`[SentenceBuilder] feed: hand=${handDetected}, gesture=${gestureState}, currentGesture=${this.currentGesture}, word="${this.currentWord}"`);

        if (!handDetected) {
            // Timer-based space/enter
            const noHandDuration = now - this.lastHandTime;
            if (noHandDuration >= this.enterNoHandMs && this.currentWord) {
                this.completeSentence();
            } else if (noHandDuration >= this.spaceNoHandMs && this.currentWord) {
                this.insertSpace();
            }
            // Reset gesture tracking when no hand
            this.currentGesture = null;
            this.gestureTriggered = false;
            return;
        }

        this.lastHandTime = now;

        // Track gesture hold time
        if (gestureState === 'fist' || gestureState === 'palm') {
            if (gestureState !== this.currentGesture) {
                // New gesture started
                console.log(`[SentenceBuilder] New gesture started: ${gestureState}`);
                this.currentGesture = gestureState;
                this.gestureStartTime = now;
                this.gestureTriggered = false;
            } else {
                // Same gesture continuing
                const gestureDuration = now - this.gestureStartTime;
                console.log(`[SentenceBuilder] Gesture hold: ${gestureState} for ${gestureDuration}ms (need ${this.gestureHoldMs}ms), triggered=${this.gestureTriggered}`);

                // Trigger action after hold time
                if (gestureDuration >= this.gestureHoldMs && !this.gestureTriggered) {
                    this.gestureTriggered = true;
                    console.log(`[SentenceBuilder] Gesture triggered! Duration: ${gestureDuration}ms`);

                    if (gestureState === 'fist') {
                        // Fist hold = backspace (delete last letter)
                        console.log(`[SentenceBuilder] Calling backspace(), currentWord="${this.currentWord}"`);
                        this.backspace();
                        this.onGesture('backspace');
                        this.onBackspace(this.currentWord);
                        console.log(`[SentenceBuilder] After backspace: currentWord="${this.currentWord}"`);
                    } else if (gestureState === 'palm') {
                        // Palm hold = delete word
                        console.log(`[SentenceBuilder] Calling deleteWord(), currentWord="${this.currentWord}"`);
                        this.deleteWord();
                        this.onGesture('delete_word');
                        this.onDeleteWord(this.currentWord);
                        console.log(`[SentenceBuilder] After deleteWord: currentWord="${this.currentWord}"`);
                    }
                }
            }
            return;
        }

        // Reset gesture tracking if not fist/palm
        if (gestureState !== this.currentGesture) {
            this.currentGesture = null;
            this.gestureTriggered = false;
        }

        // Letter stability check
        if (prediction && prediction.confidence > 0.5) {
            const letter = prediction.letter;
            if (letter === this.currentLetter) {
                // Same letter - check stability
                if (!this.letterStable && (now - this.letterStartTime >= this.stabilityThreshold)) {
                    this.letterStable = true;
                    this.commitLetter(letter);
                }
            } else {
                // New letter - restart timer
                this.currentLetter = letter;
                this.letterStartTime = now;
                this.letterStable = false;
            }
        }
    }

    /**
     * Commit a letter to the current word (with deduplication).
     * @param {string} letter
     */
    commitLetter(letter) {
        if (letter === this.lastCommittedLetter) return; // dedup
        this.currentWord += letter;
        this.lastCommittedLetter = letter;
        this.onLetterCommit(letter, this.currentWord);
    }

    /** Insert a space, finalizing the current word. */
    insertSpace() {
        if (!this.currentWord) return;
        this.words.push(this.currentWord);
        this.onWordComplete(this.currentWord, this.words);
        this.currentWord = '';
        this.lastCommittedLetter = '';
        this.currentLetter = null;
    }

    /** Complete the current sentence from accumulated words. */
    completeSentence() {
        if (this.currentWord) this.insertSpace();
        if (this.words.length === 0) return;
        const sentence = this.words.join(' ');
        this.sentences.push(sentence);
        this.onSentenceComplete(sentence, this.sentences);
        this.words = [];
    }

    /** Backspace - delete the last letter from current word. */
    backspace() {
        if (this.currentWord.length > 0) {
            this.currentWord = this.currentWord.slice(0, -1);
            this.lastCommittedLetter = this.currentWord.length > 0
                ? this.currentWord[this.currentWord.length - 1]
                : '';
            this.onBackspace(this.currentWord);
        }
    }

    /** Delete the entire current word. */
    deleteWord() {
        if (this.currentWord.length > 0) {
            const deletedWord = this.currentWord;
            this.currentWord = '';
            this.lastCommittedLetter = '';
            this.currentLetter = null;
            this.onDeleteWord(deletedWord);
        }
    }

    // Manual triggers (for UI buttons)

    manualSpace() {
        this.insertSpace();
    }

    manualEnter() {
        this.completeSentence();
    }

    clearAll() {
        this.currentWord = '';
        this.words = [];
        this.lastCommittedLetter = '';
        this.currentLetter = null;
    }

    /** @returns {Object} snapshot of current builder state */
    getState() {
        return {
            currentWord: this.currentWord,
            words: [...this.words],
            sentences: [...this.sentences],
            buildingSentence: [...this.words, this.currentWord].filter(Boolean).join(' '),
        };
    }
}

// Export for use in different contexts
if (typeof module !== 'undefined' && module.exports) {
    module.exports = { SentenceBuilder };
} else {
    window.SentenceBuilder = SentenceBuilder;
}
